packages/appcd-nodejs/src/nodejs.js

```diff
@@ -386,66 +402,157 @@ export function spawnNode({ arch, args, detached, nodeHome, nodeArgs, stdio, v8m
+				if (detached) {
+					child.unref();
 				}
-			}
+				return child;
+			})
+			.catch(err => {
+				if ((err.code === 'ETXTBSY' || err.code === 'EBUSY') && tries) {
+					logger.log(`Spawn threw ${err.code}, retrying...`);
+					return new Promise((resolve, reject) => {
+						setTimeout(() => trySpawn().then(resolve, reject), 50);
+					});
+				}
+				throw err;
+			});
+	}());
+}
+
+/**
+ * Finds all unused Node.js executables that are older than the specified max age and deletes them.
+ *
+ * @param {Number} maxAge - The maximum age for unused Node.js executables.
+ * @returns {Array.<String>} A list of unused Node.js executables that were removed.
+ */
+export function purgeUnusedNodejsExecutables({ maxAge, nodeHome }) {
+	const purged = [];
+
+	if (!isDir(nodeHome)) {
+		return purged;
+	}
 
-			args.unshift.apply(args, nodeArgs);
+	const nodeExecutable = getNodeFilename();
+	const now = Date.now();
 
-			const opts = {
-				stdio: 'inherit'
-			};
+	const isEmpty = dir => {
+		return fs.readdirSync(dir).filter(name => name === '.DS_Store').length > 0;
+	};
 
-			if (detached) {
-				opts.detached    = true;
-				opts.stdio       = 'ignore';
-				opts.windowsHide = true;
-			}
+	for (const version of fs.readdirSync(nodeHome)) {
+		const verDir = path.join(nodeHome, version);
+		if (!isDir(verDir)) {
+			continue;
+		}
 
-			if (stdio) {
-				// if stdio is set, then override the default
-				opts.stdio = stdio;
+		for (const platform of fs.readdirSync(verDir)) {
+			const platformDir = path.join(verDir, platform);
+			if (!isDir(platformDir)) {
+				continue;
 			}
 
-			let prettyArgs = args
-				.map(s => {
-					return s.indexOf(' ') === -1 ? s : `"${s}"`;
-				})
-				.join(' ');
+			for (const arch of fs.readdirSync(platformDir)) {
+				const archDir = path.join(platformDir, arch);
 
-			logger.log('Spawning: %s', highlight(`${node} ${prettyArgs} # ${JSON.stringify(opts)}`));
+				if (!isDir(archDir)) {
+					continue;
+				}
 
-			let tries = 3;
+				const nodePath = path.join(archDir, nodeExecutable);
+				if (!isFile(nodePath)) {
+					continue;
+				}
 
-			return (function trySpawn() {
-				return Promise.resolve()
-					.then(() => {
-						tries--;
-						const child = spawn(node, args, opts);
-						if (detached) {
-							child.unref();
-						}
-						return child;
-					})
-					.catch(err => {
-						if ((err.code === 'ETXTBSY' || err.code === 'EBUSY') && tries) {
-							logger.log(`Spawn threw ${err.code}, retrying...`);
-							return new Promise((resolve, reject) => {
-								setTimeout(() => trySpawn().then(resolve, reject), 50);
+				const lastRunFile = path.join(archDir, '.lastrun');
+
+				if (isFile(lastRunFile)) {
+					const lastRun = parseInt(fs.readFileSync(lastRunFile, 'utf8').trim());
+					if (!isNaN(lastRun)) {
+						if ((lastRun + maxAge) < now) {
+							purged.push({
+								version,
+								platform,
+								arch
 							});
+
+							fs.removeSync(archDir);
```

Comment:
Is there a reason to only remove the archDir initially? The verDir/platformDir stick around until the next run through of the cleanup job. Not an issue but just a question

Reply:
We remove the `archDir` so that it nukes the node executable and `.lastrun` file in one shot.

After the looping over all the arch directories, it will clean up empty platform and empty version directories.